### PR TITLE
[PSUPCLPL-12244] HAProxy Maintenance Mode

### PR DESCRIPTION
**Pull request by pranavcracker**
### Description
Adding a new procedure to Kubemarine distribution for managing the maintenance mode for HAProxy load balancers.

Fixes # (issue)
https://github.com/Netcracker/KubeMarine/issues/269 

### Solution

* We are developing a new procedure which will help us to enable/disable maintenance mode by a single command as below

### How to apply
Once the changes mentioned in this PR, we will have a automated procedure to enable or disable the maintenance mode on HAProxy load balancers. 
* Go to DVM --> Maintenance Job --> Build with parameters --> Select "haproxy_mntc" procedure from the drop down list --> provide the required argument (--mode enable/disable) --> Run the job
* Manual step -
Login to CLI and navigate to kubemarine directory --> run the following command
```
kubemarine haproxy_mntc --mode enable/disable
```

### Test Cases
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

**TestCase 1**

Run **haproxy_mntc** procedure for the cluster where maintenance mode is not supported.
`kubemarine haproxy_mntc --mode enable`
ER- procedure should fail with error message "Skiping haproxy maintenance procedure as cluster is not supported for HAProxy Maintenance mode."

**TestCase 2**

Run **haproxy_mntc** procedure for the cluster where maintenance mode is supported.

1. `kubemarine haproxy_mntc --mode enable`
2. `kubemarine haproxy_mntc --mode disable`

ER- Procedure should succeed  with enabling or disabling the maintenance mode on haproxy load balancer.

### Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] Integration CI passed
- [ ] Unit tests. If Yes list of new/changed tests with brief description
- [x] There is no merge conflicts

### CODE REVIEW

File: kubemarine/procedures/haproxy_mntc.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import os
+from collections import OrderedDict
+import time
+from kubemarine.core import flow
+from kubemarine.core.action import Action
+from kubemarine.core.resources import DynamicResources
+from kubemarine.haproxy import is_maintenance_mode
+
+def haproxy_mntc(cluster):
```
Comment (ilia1243) [inline kubemarine/procedures/haproxy_mntc.py:23]: I suggest another approach, which is more complex but more Kubemarine-like in my opinion.
1. Introduce new inventory
   ```yaml
   services:
     loadbalancer:
       haproxy:
         maintenance_mode:
           install: False
           enabled: False
           config_location: /etc/haproxy/haproxy_mntc.cfg
           keep_updated: True
   ```
   The above are default values (defaults.yaml)
   install - whether to install maintenance config haproxy_mntc.cfg. Has the same meaning as the `services.loadbalancer.haproxy.maintenance_mode` currently.
   enabled - whether to enable the maintenance mode on balancer nodes.
   config_location - moved from `services.loadbalancer.haproxy.mntc_config_location`
   keep_updated - If false, `kubemarine install --tasks deploy.loadbalancer.haproxy.configure` will not switch mode (see p. 2 below).
2. `install --tasks deploy.loadbalancer.haproxy.configure` configures and installs both configs if `services.loadbalancer.haproxy.keep_configs_updated == True` and switches mode if `maintenance_mode.keep_updated == True`
3. add/remove procedures only configure and install both configs if `services.loadbalancer.haproxy.keep_configs_updated == True`
4. New `haproxy_mntc` is not necessary in this approach, but if implemented, it only switches mode ignoring value of `maintenance_mode.keep_updated`. It has procedure inventory of the similar format as in p. 1 but supporting only `maintenance_mode.enabled`.
5. Write patch (see [Kubemarine Migration Procedure](https://github.com/Netcracker/KubeMarine/blob/main/documentation/Maintenance.md#kubemarine-migration-procedure) that changes format of `haproxy.maintenance_mode` in inventory, and sets `maintenance_mode.keep_updated=False` for backward compatibility.

Comment (pranavcracker) [reply kubemarine/procedures/haproxy_mntc.py:23]: Hi,
Yes, the suggested approach is more complex.
But the approach used in this PR is simpler and more effective as well.
We are able to enable/disable maintenance mode on haproxy with ease. 
If maintenance mode is not supported by the cluster this procedure will not work in that cluster and will give the appropriate error message.

Comment (koryaga) [reply kubemarine/procedures/haproxy_mntc.py:23]: Let's have an arch discussion about this topic. Until then - postponed.

Comment (koryaga) [reply kubemarine/procedures/haproxy_mntc.py:23]: Decided to cancel current implementation in the favor of provided above proposal. The further implementation should be start with detailed ADR and based on this comment

**Comment by koryaga** [comment]
decided to redesign

**Pull request closed without merging**